addons/udes_stock/utils.py

```diff
@@ -0,0 +1,102 @@
+from urllib.parse import urljoin
+from itertools import groupby
+from odoo import _
+
+
+def location_to_dict(location, **extras):
+    """Represent a location in a format that json can understand"""
+    if location.u_blocked:
+        reason = location.u_blocked_reason or "no reason specified"
+    else:
+        reason = False
+
+    loc = {
+        "barcode": location.barcode,
+        "name": location.name,
+        "blocked": {"is_blocked": location.u_blocked, "reason": reason},
+        "id": location.id,
+    }
+    loc.update(extras)
+    return loc
+
+
+def product_to_dict(product, **extras):
+    """Represent a product in a format that json can understand"""
+    prod = {"name": product.name, "barcode": product.barcode, "display_name": product.display_name}
+    prod.update(extras)
+    return prod
+
+
+def format_picking_data_for_display_list_component(picking, order=True):
+    """Formats the data of a picking to be displayed in a list component
+
+    The data will be returned as a list of strings "packagename productname x productquantity"
+    """
+    contents = []
+    for package, package_lines in picking.move_line_ids.groupby("package_id", order):
+        for product, product_lines in package_lines.groupby("product_id", order):
+            contents.append(
+                package_product_quantity_label(
+                    package, product, int(sum(product_lines.mapped("product_uom_qty")))
+                )
+            )
+    return contents
+
+
+def product_quantity_label(product, quantity):
+    """A function for a common output format for product quantity labels"""
+    return "{name} x {quantity}".format(name=product.display_name, quantity=quantity)
+
+
+def package_product_quantity_label(package, product, quantity):
+    """A function for a common output format for package and product quantity labels"""
+    return "{package} {product_quantity}".format(
+        package=package.name, product_quantity=product_quantity_label(product, quantity)
+    )
+
+
+def batch_to_dict(batch):
+    """Represent a batch in a format that json can understand"""
+    return {"id": batch.id, "name": batch.u_original_name or batch.name}
+
+
+def sorted_group_by(iterable, key):
+    """Sorts and then groups an iterable by a key"""
+    return groupby(sorted(iterable, key=key), key=key)
+
+
+def product_image_urls(base_url, product):
+    """Produces a product image urls object"""
+    if product.image:
+        base_url = urljoin(base_url, "/web/image/product.product/%i" % product.id)
```

Comment:
Might be good idea to import this path from BASE_PRODUCT_IMAGE_URL in "addons/udes_stock/models/product_product.py" so that it's not a magic value as in appearing from nowhere.

Reply:
Yep good suggestion, made the changes and pushed them